torchdynamo/convert_frame.py

```diff
@@ -101,6 +103,52 @@ def _fn(*args, **kwargs):
     return _fn
 
 
+def has_tensor_in_frame(frame):
+    """Check if the frame has torch.* related bits"""
+    # Check if the function was decorated with torchdynamo.optimize
+    if frame.f_code in always_optimize_code_objects:
+        return True
+
+    # Check if there is global import of torch.*
+    for co_name in frame.f_code.co_names:
+        if co_name in frame.f_globals:
+            if is_allowed(frame.f_globals[co_name]):
+                return True
+
+    def has_tensor(obj):
+        """Recursively check if the obj has a tensor"""
+        if isinstance(obj, (torch.Tensor, torch.nn.Module)):
+            return True
+        elif isinstance(obj, (list, tuple)):
```

Comment:
I think it might be possible to call into user code here.
```
class Foo(list):
   def __iter__(self):
       raise Exception()
```

We could use `utils.istype` for a more strict type check.  We could call `list.values(obj)` directly.  

Similar issues below.  We should be careful about user defined objects.